modular_skyrat/modules/reagent_forging/code/forge.dm

```diff
@@ -80,213 +139,233 @@
 		var/obj/item/stack/sheet/stack_sheet = I
 		if(!stack_sheet.use(1)) //need to be able to use the item, so no glue
-			to_chat(user, span_warning("You abandon lining the forge with sinew."))
-			in_use = FALSE
+			fail_message(user, "You abandon lining [src] with [I].")
 			return
 		playsound(src, 'sound/magic/demon_consume.ogg', 50, TRUE)
 		sinew_lower_chance += 10
 		current_sinew++
 		in_use = FALSE
-		to_chat(user, span_notice("You successfully line the forge with sinew."))
+		to_chat(user, span_notice("You successfully line [src] with [I]."))
 		return
 
 	if(istype(I, /obj/item/organ/regenerative_core))
+		var/skill_modifier = user.mind.get_skill_modifier(/datum/skill/smithing, SKILL_SPEED_MODIFIER) * 3 SECONDS
 		if(in_use) //only insert one at a time
 			to_chat(user, span_warning("You cannot do multiple things at the same time!"))
 			return
 		in_use = TRUE
 		if(reagent_forging) //if its already able to reagent forge, why continue wasting?
-			to_chat(user, span_warning("This forge is already upgraded."))
-			in_use = FALSE
+			fail_message(user, "[src] is already upgraded.")
 			return
 		var/obj/item/organ/regenerative_core/used_core = I
 		if(used_core.inert) //no inert cores allowed
-			to_chat(user, span_warning("You cannot use an inert regenerative core."))
-			in_use = FALSE
+			fail_message(user, "You cannot use an inert [used_core].")
 			return
-		to_chat(user, span_warning("You start to sacrifice the regenerative core to the forge..."))
-		if(!do_after(user, 3 SECONDS, target = src)) //wait 3 seconds to upgrade
-			to_chat(user, span_warning("You abandon sacrificing the regenerative core to the forge."))
-			in_use = FALSE
+		to_chat(user, span_warning("You start to sacrifice [used_core] to [src]..."))
+		if(!do_after(user, skill_modifier, target = src)) //wait 3 seconds to upgrade
+			fail_message(user, "You abandon sacrificing [used_core] to [src].")
 			return
-		to_chat(user, span_notice("You successfully sacrifice the regenerative core to the forge."))
+		to_chat(user, span_notice("You successfully sacrifice [used_core] to [src]."))
 		playsound(src, 'sound/magic/demon_consume.ogg', 50, TRUE)
 		qdel(I)
 		current_core++
 		in_use = FALSE
 		if(current_core >= 3) //use three regenerative cores to get reagent forging capabilities on the forge
-			reagent_forging = TRUE
-			to_chat(user, span_notice("You feel the forge has upgraded."))
-			color = "#ff5151"
-			name = "reagent forge"
-			desc = "A structure built out of metal, with the intended purpose of heating up metal. It has the ability to imbue!"
+			create_reagent_forge()
 		return
 
 	if(istype(I, /obj/item/stack/sheet/animalhide/goliath_hide))
+		var/skill_modifier = user.mind.get_skill_modifier(/datum/skill/smithing, SKILL_SPEED_MODIFIER) * 6 SECONDS
 		var/obj/item/stack/sheet/animalhide/goliath_hide/goliath_hide = I
 		if(in_use) //only insert one at a time
 			to_chat(user, span_warning("You cannot do multiple things at the same time!"))
 			return
```

Comment:
honestly, putting this as the FIRST if() in the whole proc will probably do fine; functionally checking 'are we being used' before checking anything else